### PR TITLE
WIP: Add snirf writer (moved)

### PR DESCRIPTION
**Pull request by rob-luke**
Moved from #146 to my fork.

Add support for writing SNIRF files.

mne-tools/mne-python#8060

### CODE REVIEW

**Comment by rob-luke** [comment]
These tests pass locally with the branch from https://github.com/mne-tools/mne-python/pull/8079

**Comment by rob-luke** [comment]
@agramfort and @larsoner could you please review again. I would appreciate feedback if you think the method of saving `raw.info['dig']` is appropriate. Could you please review in parallel with https://github.com/mne-tools/mne-python/pull/8079

File: mne_nirs/io/snirf/tests/test_snirf.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    # Check info object is the same
+    obj_diff = object_diff(raw.info, raw_orig.info)
+    num_diffs = 0
+    for line in obj_diff.splitlines():
+        if ('logno' not in line) and ('scanno' not in line):
+            # logno and scanno are not used in processing
+            num_diffs += 1
+            print(line)
+    assert num_diffs == 0
```
Comment (larsoner) [inline mne_nirs/io/snirf/tests/test_snirf.py:53]: I would do this instead so it shows up directly in the error:
```suggestion
    diffs = ''
    for line in obj_diff.splitlines():
        if ('logno' not in line) and ('scanno' not in line):
            # logno and scanno are not used in processing
            diffs += f'\n{line}'
    assert diffs == ''
```
But I'm a bit surprised that the scanno and logno actually change. Seems like those should stay the same

File: mne_nirs/io/snirf/_snirf.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
+# Authors: Robert Luke <mail@robertluke.net>
+#
+# License: BSD (3-clause)
+
+import h5py as h5py
+from re import match as mch
```
Comment (agramfort) [inline mne_nirs/io/snirf/_snirf.py:6]: I would not use this custom shortcut. It required me to search in the file to know what `mch` was. I would use re.match to be explicit. It's not is too many places anyway

**Pull request merged**